src/parser/block-additions/core-block.php

```diff
@@ -88,105 +52,47 @@ public static function cleanup_after_render() {
+	 * @param int|null $post_id      Post ID.
 	 * @param array    $parsed_block Parsed block data.
 	 * @return array
 	 */
-	public static function get_inner_blocks( array $inner_blocks, string $block_name, int|null $_post_id, array $parsed_block ): array {
+	public static function get_inner_blocks( array $inner_blocks, string $block_name, int|null $post_id, array $parsed_block ): array {
+		// Not a synced pattern? Return the inner blocks unchanged.
 		if ( self::$block_name !== $block_name || ! isset( $parsed_block['attrs']['ref'] ) ) {
 			return $inner_blocks;
 		}
 
-		$store_key = self::get_store_key( $parsed_block );
+		$context = [];
 
-		if ( ! isset( self::$captured_inner_blocks[ $store_key ] ) ) {
-			return $inner_blocks;
+		// Support synced pattern overrides. Copied and adapted from core:
+		// https://github.com/WordPress/WordPress/blob/6.6.1/wp-includes/blocks/block.php#L81
+		//
+		// In our case, we don't need to filter the context since we can pass it in.
+		if ( isset( $parsed_block['attrs']['content'] ) ) {
+			$context['pattern/overrides'] = $parsed_block['attrs']['content'];
 		}
 
-		return self::$captured_inner_blocks[ $store_key ]['inner_blocks'];
-	}
+		// Load, parse, and render the inner blocks of the synced pattern, passing
+		// along its block context. We intentionally do not recursively call
+		// ContentParser->parse() to avoid calling telemetry and filters again.
+		$parser = new ContentParser();
+		$post   = get_post( $parsed_block['attrs']['ref'] );
 
-	/**
-	 * Create a unique key that can be used to identify a synced pattern. This
-	 * allows us to store and retrieve inner blocks for synced patterns and avoid
-	 * duplication when they are used multiple times within the same tree.
-	 *
-	 * Using a hash of attributes is important because they may contain synced
-	 * pattern overrides, which can change the inner block content. The attributes
-	 * contain the synced pattern post ID, so uniqueness is built-in.
-	 *
-	 * @param array $parsed_block Parsed block data.
-	 * @return string
-	 */
-	protected static function get_store_key( array $parsed_block ): string {
-		// Include the synced pattern ID in the key just for legibility.
-		$synced_pattern_id = $parsed_block['attrs']['ref'] ?? null;
-		$attribute_json    = wp_json_encode( $parsed_block['attrs'] );
-
-		return sprintf( '%s_%s', strval( $synced_pattern_id ), sha1( $attribute_json ) );
-	}
-
-	/**
-	 * Capture inner block for a synced pattern.
-	 *
-	 * @param array    $synced_pattern Synced pattern block (parsed block).
-	 * @param WP_Block $block          Inner block.
-	 */
-	protected static function capture_inner_block( array $synced_pattern, WP_Block $block ): void {
-		$store_key = self::get_store_key( $synced_pattern );
-		if ( ! isset( self::$captured_inner_blocks[ $store_key ] ) ) {
-			self::$captured_inner_blocks[ $store_key ] = [
-				'inner_blocks' => [],
-				'locked'       => false,
-			];
+		if ( ! $post instanceof \WP_Post ) {
```

Comment:
Extreme nit - can we use `use WP_Block;` up top and this for consistency?

```suggestion
		if ( ! $post instanceof WP_Post ) {
```

Reply:
Nice catch! Fixed in https://github.com/Automattic/vip-block-data-api/pull/87/commits/fc2457531a99621d2d8ca62c1a2728df8f0e7fbf